sockeye/data_io.py

```diff
@@ -357,7 +361,67 @@ def shard_data(source_fnames: List[str],
+                sample_scores = [int(line.strip()) for line in curriculum_score_file.readlines() if line.strip()]
+
+            # sample_shard_mapping: a list stores the mapping relation between sample and shard
+            # index is the index of sample
+            # each element is the shard the sentence will be put into
+            sample_shard_mapping = list(sample_scores)
+
+            # shard_score_mapping: a list stores the mapping relation between shards and complexity scores
+            # index is the index of shard
+            # each element is the complexity score of the shard
+            shard_score_mapping = list(range(num_shards))
+
+            # unique_scores: a sorted list stores unique complexity scores
+            unique_scores = sorted(set(sample_scores))
+
+            # score_index_dict: {score 0: [sample_index...], score 1: [sample_index..]...}
+            score_sample_dict = defaultdict(list)
+            for i in range(len(sample_scores)):
+                score_sample_dict[sample_scores[i]].append(i) 
+
+
+            if num_shards > len(unique_scores):
+                shard_count = 0
+                # if num_shards is multiple of number of unique scores,
+                # samples with different complexity scores should be put into shards of equal numbers
+                if num_shards%len(unique_scores) == 0:
+                    num_shards_per_score = int(num_shards/len(unique_scores))         
+                    for us in unique_scores:
+                        for sublist in np.array_split(np.array(score_sample_dict[us]),num_shards_per_score):
+                            for index in sublist:
+                                sample_shard_mapping[index] = shard_count 
+                            shard_score_mapping[shard_count] = us
+                            shard_count += 1
+
+                # if num_shards is not multiple of number of unique scores,
+                # samples will be allocated according to samples_per_shard
+                else:
+                    for us in unique_scores:
+                        sample_count = 0
+                        for index in sorted(score_sample_dict[us]):
+                            sample_shard_mapping[index] = shard_count
+                            sample_count += 1
+                            if sample_count==samples_per_shard:
+                                shard_score_mapping[shard_count] = us
+                                shard_count += 1
+                                sample_count = 0
+                        if sample_count < samples_per_shard:
+                            shard_score_mapping[shard_count] = us
+                            shard_count += 1
+                                
+
+            with open(os.path.join(output_prefix, 'shard_scores'), 'w') as shard_scores_file:
+                for m in shard_score_mapping:
+                    shard_scores_file.write(str(m)+'\n')
```

Comment:
spacing.
Since this is Python3, you could use print(m, file=shard_scores_file) here. newline will be appended by print()